File: solutions/typescript/0010.ts
function isMatch(s: string, p: string): boolean {
    const m = s.length;
    const n = p.length;

    const dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(false));
    dp[0][0] = true;

    for (let i = 0; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (p[j - 1] === '*') {
                dp[i][j] = dp[i][j - 2] || (i > 0 && (s[i - 1] === p[j - 2] || p[j - 2] === '.') && dp[i - 1][j]);
            } else {
                dp[i][j] = i > 0 && (s[i - 1] === p[j - 1] || p[j - 1] === '.') && dp[i - 1][j - 1];
            }
        }
    }

    return dp[m][n];
}

File: solutions/rust/0010.rs
impl Solution {
    pub fn is_match(s: String, p: String) -> bool {
        let m = s.len();
        let n = p.len();

        let mut dp = vec![vec![false; n + 1]; m + 1];
        dp[0][0] = true;

        for i in 0..=m {
            for j in 1..=n {
                if p.chars().nth(j - 1).unwrap() == '*' {
                    dp[i][j] = dp[i][j - 2] || (i > 0 && (s.chars().nth(i - 1).unwrap() == p.chars().nth(j - 2).unwrap() || p.chars().nth(j - 2).unwrap() == '.') && dp[i - 1][j]);
                } else {
                    dp[i][j] = i > 0 && (s.chars().nth(i - 1).unwrap() == p.chars().nth(j - 1).unwrap() || p.chars().nth(j - 1).unwrap() == '.') && dp[i - 1][j - 1];
                }
            }
        }
        dp[m][n]
    }
}